File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int solve(int ind, int size, vector<int>& nums){
        if(ind == size)return 0;

        int prePrev = 0;
        int prev = nums[ind];

        for(int i = ind + 1; i < size; i++){
            int take = prePrev + nums[i];
            int skip = prev;

            int temp = max(take, skip);

            prePrev = prev;
            prev = temp;
        }

        return prev;
    }

    int rob(vector<int>& nums) {
        if(nums.size() == 1) return nums[0];
        if(nums.size() == 2) return max(nums[0], nums[1]);

        return max(solve(0, nums.size()-1, nums),
                    solve(1, nums.size(), nums));
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;

        if (n == 0) return 0;
        if (n == 1) return nums[0];

        List<Integer> arr1 = new ArrayList<>();
        List<Integer> arr2 = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            if (i != 0) arr1.add(nums[i]);
            if (i != n - 1) arr2.add(nums[i]);
        }

        return Math.max(solve(arr1), solve(arr2));
    }


    public int solve(List<Integer> arr) {
        int m = arr.size();
        int[] house = new int[m + 1];

        house[0] = 0;
        house[1] = arr.get(0);

        for(int i = 2; i <= m; i++){
            int take = arr.get(i - 1) + house[i - 2];
            int notTake = house[i - 1];

            house[i] = Math.max(take, notTake);
        }

        return house[m];
    }
}